docs/proposals/parameterized-config-management-plugins.md

```diff
@@ -0,0 +1,502 @@
+    1. It must present an equivalent parameters UI.
+    2. It must communicate errors back to the repo-server (and then the UI) the same as the native implementation.
+
+### Non-Goals
+
+We should not:
+* Re-implement config management tools as CMPs (besides Helm)
+
+## Proposal
+
+### Use cases
+
+#### Use case 1: building Argo CD without config management dependencies
+
+As an Argo CD developer, I would like to be able to build Argo CD without including the Helm SDK as a dependency.
+
+The Helm SDK includes the Kubernetes code base. That's a lot of code, and it will make builds unacceptably slow.
+
+#### Use case 2: writing CMPs with rich UI experiences
+
+As an Argo CD user, I would like to be able to parameterize manifests built by a CMP.
+
+For example, if the Argo CD administrator has installed a CMP which applies a last-mile kustomize overlay to a Helm
+repo, I would like to be able to pass values to the Helm chart without having to manually discover those parameter names
+(in other words, they should show up in the Application UI just like with a native Helm Application). I also shouldn't 
+have to ask my Argo CD admin to modify the CMP to accommodate the values as environment variables.
+
+### Implementation Details/Notes/Constraints
+
+#### Prerequisites
+
+Since this proposal is designed to increase CMP adoption, we need to make sure there aren't any bugs that make CMPs
+less robust than native tools.
+
+Bugs to fix:
+1. [#8145](https://github.com/argoproj/argo-cd/issues/8145) - `argocd app sync/diff --local` doesn't account for sidecar CMPs
+
+#### Terms
+
+* **Parameter definition**: an instance of a data structure which describes an individual parameter that may be applied
+  to a specific Application. (See the [schema](#parameter-definition-schema) below.)
+* **Parameters announcement**: a list of parameter definitions. (See the [schema](#parameters-announcement-schema) below.)
+
+  "Parameters" is plural because each "announcement" will be a list of multiple parameter definitions.
+* **Parameterized CMP**: a CMP which supports rich parameters (i.e. more than environment variables). A CMP is
+  parameterized if either of these is true:
+  1. its configuration includes the sections consumed by the default CMP server to generate parameters announcements
+  2. it is a fully customized CMP server which implements an endpoint to generate parameters announcements
+
+#### CMP config schema
+
+This proposal adds a new `parameters` key to the ConfigManagementPlugin config spec.
+
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: ConfigManagementPlugin
+metadata:
+  name: cmp-plugin
+spec:
+  version: v1.0
+  generate:
+    command: ["example.sh"]
+  discover:
+    fileName: "./subdir/s*.yaml"
+  # NEW KEY
+  parameters:
+    # The declarative announcement follows the parameters announcement schema. This is where a parameter description
+    # should go if it applies to all apps for this CMP.
+    announcement:
```

Comment:
It seems to me that announcement is more an implementation concern and Im not sure if it should be part of this contract. 

Why not having something simpler like:

```yaml
  parameters:
  - name: example
  dynamicParameters:
  - command: ["example-params.sh"]
```

In this case both would be merged and announced to the UI.

Reply:
Yep, that's a lot simpler. Will change the proposal.